Sources/ScipioKit/Producer/FrameworkProducer.swift

```diff
@@ -141,22 +120,93 @@ struct FrameworkProducer {
         }
     }
 
-    private func restoreAllAvailableCaches(
+    private func restoreAllAvailableCachesIfNeeded(
         availableTargets: Set<CacheSystem.CacheTarget>,
         cacheSystem: CacheSystem
     ) async -> Set<CacheSystem.CacheTarget> {
-        let chunked = availableTargets.chunks(ofCount: cacheStorage?.parallelNumber ?? CacheSystem.defaultParalellNumber)
+        let cacheStorages: [any CacheStorage]
+
+        switch cacheMode {
+        case .disabled:
+            return []
+        case .project:
+            // For `.project`, just checking whether the valid caches (already built frameworks under the project)
+            // exist or not (not restoring anything from external locations).
+            return await restoreCachesForTargets(
+                availableTargets,
+                cacheSystem: cacheSystem,
+                cacheStorage: nil
+            )
+        case .storage(let storage, let actors):
+            guard actors.contains(.consumer) else { return [] }
+            cacheStorages = [storage]
+        case .storages(let storages):
+            let storagesWithConsumer = storages.compactMap { storage, actors in
+                actors.contains(.consumer) ? storage : nil
+            }
+            guard !storagesWithConsumer.isEmpty else { return [] }
+            cacheStorages = storagesWithConsumer
+        }
+
+        var remainingTargets = availableTargets
+        var restored: Set<CacheSystem.CacheTarget> = []
+
+        for index in cacheStorages.indices {
+            let storage = cacheStorages[index]
+
+            let logSuffix = "[\(index)] \(type(of: storage))"
+            if index == cacheStorages.startIndex {
+                logger.info(
+                    "▶️ Starting restoration with cache storage: \(logSuffix)",
+                    metadata: .color(.green)
+                )
+            } else {
+                logger.info(
+                    "⏭️ Falling back to next cache storage: \(logSuffix)",
+                    metadata: .color(.green)
+                )
+            }
+
+            let restoredPerStorage = await restoreCachesForTargets(
+                remainingTargets,
+                cacheSystem: cacheSystem,
+                cacheStorage: storage
+            )
+            restored.formUnion(restoredPerStorage)
+
+            logger.info(
+                "⏸️ Restoration finished with cache storage: \(logSuffix)",
+                metadata: .color(.green)
+            )
+
+            remainingTargets.subtract(restoredPerStorage)
+            if remainingTargets.isEmpty {
+                break
+            }
```

Comment:
if all frameworks are successfully restored, we don't need to proceed to next cache storage.

Reply:
[not mandatory] it's kindly to comment this